gemini-jax-rs/src/main/java/com/techempower/gemini/jaxrs/core/JaxRsDispatcher.java

```diff
@@ -0,0 +1,911 @@
+    String[] segments = uri.split("/");
+    return rootBlock.getDispatchMatch(httpMethod, segments, 0);
+  }
+
+  public DispatchBestMatchInfo getBestMatchInfo(DispatchMatchSource dispatchMatchSource) {
+    String uri = dispatchMatchSource.uri;
+    String httpMethod = dispatchMatchSource.httpMethod;
+    int uriStart = uri.startsWith("/") ? 1 : 0;
+    int uriEnd = uri.length() - (uri.endsWith("/") ? 1 : 0);
+    String normalizedUri = uri.substring(uriStart, uriEnd);
+    DispatchMatch dispatchMatch = getDispatchMatches(httpMethod, normalizedUri);
+    DispatchBestMatchInfo matchInfo = new DispatchBestMatchInfo();
+    // TODO: Need to filter the dispatch match down to only the correct one so
+    //  the path params can be extracted. There's a TO-DO...somewhere in this
+    //  file (or maybe the test) for how to do that naturally and only come out
+    //  with the right match per the sorting rules, and not have to deal with
+    //  the tree of matches.
+    dispatchMatch.getBestMatch(matchInfo);
+    return matchInfo;
+  }
+
+  public DispatchBestMatchInfo getBestMatchInfo(String httpMethod, String uri) {
+    DispatchMatchSource matchSource = new DispatchMatchSource(httpMethod, uri);
+    if (cacheEnabled) {
+      if (useResinCache) {
+        DispatchBestMatchInfo matchInfo = bestMatchesResinCache.get(matchSource);
+        if (matchInfo == null) {
+          matchInfo = getBestMatchInfo(matchSource);
+          bestMatchesResinCache.put(matchSource, matchInfo);
+        }
+        return matchInfo;
+      } else if (useCaffCache) {
+        return bestMatchesCaff.get(matchSource);
+      } else {
+        try
+        {
+          return bestMatches.get(matchSource);
+        }
+        catch (ExecutionException e)
+        {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+    return getBestMatchInfo(matchSource);
+  }
+
+  // TODO: Not use Gemini's Context eventually, probably.
+  //public void dispatch(Context context)
+  public Object dispatch(String httpMethod, String uri)
+  {
+    // TODO: Add `getServletPath` as a method in Context so that this can
+    //       only match the URI relative to where the servlet is hosted.
+    //       `getServletPath` is a method of HttpServletRequest. For
+    //       non-servlet containers this can just default to "" or "/". Find
+    //       out what the default is for Resin and use that.
+    DispatchBestMatchInfo matchInfo = getBestMatchInfo(httpMethod, uri);
+
+    Method method = matchInfo.endpoint.method;
+    // TODO: For now I'm just gonna support valueOf for simplicity,
+    //  and assume it's static.
+    List<Object> arguments = new ArrayList<>();
+    Object instance = matchInfo.endpoint.resource.getInstance();
+    Class<?> instanceClass = matchInfo.endpoint.resource.getInstanceClass();
+    Class<?>[] parameterTypes = method.getParameterTypes();
+    Annotation[][] parameterAnnotations = method.getParameterAnnotations();
```

Comment:
It seems to me that this reflection work should be done and cached at startup, rather than on a per-request dispatch.

Reply:
There's a TODO in here somewhere for exactly that. Atm I'm just focusing on matching the spec so I have a better idea of how to structure a lot of this. It entirely possible I'll get to that caching earlier than I expect, but I'm planning on doing that once the structure is less likely to be heavily refactored (which is definitely going to happen before this dispatcher is done).